src/npe2/io_utils.py

```diff
@@ -167,12 +171,94 @@ def _read(
 
     if plugin_name:
         raise ValueError(
-            f"Plugin {plugin_name!r} was selected to open "
+            f"Reader {plugin_name!r} was selected to open "
             + f"{paths!r}, but returned no data."
         )
     raise ValueError(f"No readers returned data for {paths!r}")
 
 
+def _get_compatible_readers_by_choice(
+    plugin_name: Union[str, None], paths: Union[str, Sequence[str]], pm: PluginManager
+):
+    """Returns compatible readers filtered by validated plugin choice.
+
+    Checks that plugin_name is an existing plugin (and command if
+    a specific contribution was passed), and that it is compatible
+    with paths. Raises ValueError if given plugin doesn't exist,
+    it is not compatible with the given paths, or no compatible
+    readers can be found for paths (if no plugin was chosen).
+
+    Parameters
+    ----------
+    plugin_name: Union[str, None]
+        name of chosen plugin, or None
+    paths: Union[str, Sequence[str]]
+        paths to read
+    pm: PluginManager
+        plugin manager instance to check for readers
+
+    Raises
+    ------
+    ValueError
+        If the given reader doesn't exist
+    ValueError
+        If there are no compatible readers
+    ValueError
+        If the given reader is not compatible
+
+    Returns
+    -------
+    compat_readers : List[ReaderContribution]
+        Compatible readers for plugin choice
+    """
+    passed_contrib = plugin_name and ("." in plugin_name)
+    compat_readers = list(pm.iter_compatible_readers(paths))
+    compat_reader_names = sorted(
+        {(rdr.command if passed_contrib else rdr.plugin_name) for rdr in compat_readers}
+    )
+    helper_error_message = (
+        f"Available readers for {paths!r} are: {compat_reader_names!r}."
+        if compat_reader_names
+        else f"No compatible readers are available for {paths!r}."
+    )
+
+    # check whether plugin even exists.
+    if plugin_name:
+        try:
+            # note that get_manifest works with a full command e.g. my-plugin.my-reader
+            pm.get_manifest(plugin_name)
+            if passed_contrib:
+                pm.get_command(plugin_name)
+        except KeyError:
+            raise ValueError(
+                f"Given reader {plugin_name!r} does not exist. {helper_error_message}"
+            ) from None
+
+    # no choice was made and there's no readers to try
+    if not plugin_name and not len(compat_reader_names):
+        raise ValueError(helper_error_message)
+
+    # user didn't make a choice and we have some readers to try, return them
+    if not plugin_name:
+        return compat_readers
+
+    # user made a choice and it exists, but it may not be a compatible reader
+    plugin, _, contrib = plugin_name.partition(".")
```

Comment:
`contrib` isn't used

```suggestion
    plugin, _, _ = plugin_name.partition(".")
```